towhee/ann_index/base.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2021 Zilliz. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+
+
+class ANNIndex(ABC):
+    """
+    ANN index base class, implements __init__, insert and search.
+    """
+
+    @abstractmethod
+    def __init__(self):
+        """
+        Init ANN index class.
+        """
+        pass
+
+    @abstractmethod
+    def insert(self, *args):
+        """
+        Insert data with `ann_insert`.
+        """
+        pass
+
+    @abstractmethod
+    def search(self, *args, **kwargs):
+        """
+        Search data with `ann_search`.
+        """
+        pass
+
+    def load(self):
+        """
+        Load the index.
+        """
+        pass
+
+    def train(self):
+        """
+        Train the index.
+        """
+        pass
+
+    def save(self):
+        """
+        Save the index.
+        """
+        pass
+
+    def __enter__(self):
```

Comment:
can we put the code of entering and exiting here?

Reply:
What do you mean, such as `load` with `__exit__`? But I think each index will have different action with `__enter__` and `__exit__`.